### PR TITLE
Update onnxruntime/trainer.py with recent huggingface change

### LINKED ISSUE #1133: AttributeError: 'ORTSeq2SeqTrainingArguments' object has no attribute 'deepspeed_plugin'
### System Info

```shell
- Python 3.8
- Cuda 11.7
- Hugging Face Dependencies:
accelerate==0.20.3
datasets==2.13.1 
evaluate==0.4.0
transformers==4.30.2
optimum==1.8.6
```

### Who can help?

@JingyaHuang @echarlaix 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Environment:
```Dockerfile
FROM mcr.microsoft.com/azureml/curated/acpt-pytorch-2.0-cuda11.7
RUN pip install accelerate datasets evaluate transformers optimum
```

Run Command: `torchrun optimum/examples/onnxruntime/training/summarization/run_summarization.py --model_name_or_path t5-small --dataset_name cnn_dailymail --dataset_config "3.0.0" --source_prefix "summarize: " --do_train --do_eval --per_device_train_batch_size=4 --per_device_eval_batch_size=4 --output_dir /tmp/ort_summarization_t5/ --overwrite_output_dir --predict_with_generate`

Error:
```
Traceback (most recent call last):
  File "src/Finetune/train_summarization_deepspeed_optum.py", line 684, in <module>
    main()
  File "src/Finetune/train_summarization_deepspeed_optum.py", line 586, in main
    trainer = ORTSeq2SeqTrainer(
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/optimum/onnxruntime/trainer.py", line 304, in __init__
    super().__init__(
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/transformers/trainer.py", line 345, in __init__
    self.create_accelerator_and_postprocess()
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/transformers/trainer.py", line 3825, in create_accelerator_and_postprocess
    deepspeed_plugin=self.args.deepspeed_plugin,
AttributeError: 'ORTSeq2SeqTrainingArguments' object has no attribute 'deepspeed_plugin'
```

### Expected behavior

This is expected to train T5 model for summarization task.

### PR DESCRIPTION
**Pull request by jingyanwangms**
# What does this PR do?
This PR updates trainer.py and trainer_args.py to keep up with changes around deepspeed usage in transformers/trainer.
This starts to break since https://github.com/huggingface/transformers/commit/a73b1d59a34ca8e30098b3e32a977a7928663559#diff-ed55888e6665791fe92cc8fc0c499da54f4ace6738551cd9a2591881cda076de

Fixes #1133 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by JingyaHuang** [comment]
related #1133

**Comment by JingyaHuang** [comment]
Hi @jingyanwangms, thanks a lot for the PR. Could you confirm which release of transformers this PR is updated to?

**Comment by jingyanwangms** [comment]
> Hi @jingyanwangms, thanks a lot for the PR. Could you confirm which release of transformers this PR is updated to?
@JingyaHuang Where can I find release information?

File: optimum/onnxruntime/trainer.py [hunk 1]
```diff
@@ -528,7 +528,8 @@ def _inner_training_loop(
         model = ORTModule(self.model)
         self.model_wrapped = model
 
-        if args.deepspeed:
+        # Note transformers/trainer.py renamed this arg to is_deepspeed_enabled
+        if args.deepspeed:
```
Comment (JingyaHuang) [inline optimum/onnxruntime/trainer.py:532]: In the latest `Trainer`, `args.deepspeed` is replaced by the attribute `self.is_deepspeed_enabled` fetched from the accelerator state.

**Review by JingyaHuang — COMMENTED** [review]
Hi @jingyanwangms,

Currently in Transformers, we are refactoring distributed training features using the Accelerate library. So it would be pretty important that we make sure all functions overwritten by `ORTTrainer` are compatible (which means that we need to apply all changes in `transformers.Trainer` instead of the bugged lines to ensure that those features can work correctly).

Do you mind applying all commits from transformers 4.27, until transformers 4.30.2 in functions overwritten by `ORTTrainer`?

**Comment by jingyanwangms** [comment]
> Hi @jingyanwangms,
> 
> Currently in Transformers, we are refactoring distributed training features using the Accelerate library. So it would be pretty important that we make sure all functions overwritten by `ORTTrainer` are compatible (which means that we need to apply all changes in `transformers.Trainer` instead of the bugged lines to ensure that those features can work correctly).
> 
> Do you mind applying all commits from transformers 4.27, until transformers 4.30.2 in functions overwritten by `ORTTrainer`?

How do I apply all commits from transformers 4.27 until 4.30.2? The latest optimum branch that I can find is v1.8.8.

Also, what are your thoughts on calling TrainingArguments.__post_init__ with super.__post_init__ instead of duplicating a lot of the logic in ORTTrainingArguments.__post_init__ ?

**Comment by jingyanwangms** [comment]
@JingyaHuang can you please explain more on "apply all commit"?

**Comment by pengwa** [comment]
> @JingyaHuang can you please explain more on "apply all commit"?

I think @JingyaHuang mean, all the changes after the file optimum duplicated the training_args, e.g. compare that old commit with latest transformer training_args, apply all changes that is irrelevant to ORT. 

If the effort is too big, maybe we can also just refactor the training_args __post_init__(), letting it just call transformers correposinding __post_init__() then do ORT specific things. 

While either way it might break some other usages, I mean other examples script might also need a change accordingly. Hopefully there is no much things we need to do.

**Comment by jingyanwangms** [comment]
> > @JingyaHuang can you please explain more on "apply all commit"?
> 
> I think @JingyaHuang mean, all the changes after the file optimum duplicated the training_args, e.g. compare that old commit with latest transformer training_args, apply all changes that is irrelevant to ORT.
> 
> If the effort is too big, maybe we can also just refactor the training_args **post_init**(), letting it just call transformers correposinding **post_init**() then do ORT specific things. @JingyaHuang do you think this is feasible, or you had any concerns on this?
> 
> While either way it might break some other usages, I mean other examples script might also need a change accordingly. Hopefully there is no much things we need to do.
I'm going check recent transformers trainers.py commits as Peng suggested. @JingyaHuang please confirm if this is what you meant

**Comment by jingyanwangms** [comment]
@JingyaHuang a side note, when will transformers 5 come out? By then we can deprecate overriding __post_init__

File: optimum/onnxruntime/trainer.py (hunk 1 above)
Comment (jingyanwangms) [reply optimum/onnxruntime/trainer.py]: Changed

File: optimum/onnxruntime/trainer.py [hunk 2]
```diff
@@ -542,20 +542,31 @@ def _inner_training_loop(
...
-            deepspeed_engine, optimizer, lr_scheduler = deepspeed_init(
-                self, num_training_steps=max_steps, resume_from_checkpoint=resume_from_checkpoint
+            optimizer, lr_scheduler = deepspeed_init(
+                self, num_training_steps=max_steps
             )
-            self.model = unwrap_model(deepspeed_engine)
-            self.model_wrapped = deepspeed_engine
-            self.deepspeed = deepspeed_engine
+            # deepspeed ckpt loading
+            if resume_from_checkpoint is not None and self.is_deepspeed_enabled:
```
Comment (pengwa) [inline optimum/onnxruntime/trainer.py:549]: can we keep the same order of https://github.com/huggingface/transformers/commit/a73b1d59a34ca8e30098b3e32a977a7928663559#diff-ed55888e6665791fe92cc8fc0c499da54f4ace6738551cd9a2591881cda076deR1779 did? 

I mean we don't need put every deepspeed related things in one code block. Following what HF trainer did will make it easier for us to catch up with later HF trainer changes.

**Pull request closed without merging**